src/nunavut/lang/cpp/support/variable_length_array.hpp

```diff
@@ -849,16 +845,743 @@ class VariableLengthArray
+            --index_;
+            return tmp;
+        }
+
+        IteratorImpl& operator+=(const difference_type n)
+        {
+            index_ = static_cast<size_type>(static_cast<difference_type>(index_) + n);
+            return *this;
+        }
+        IteratorImpl& operator-=(const difference_type n)
+        {
+            index_ = static_cast<size_type>(static_cast<difference_type>(index_) - n);
+            return *this;
+        }
+        IteratorImpl operator+(const difference_type n) const
+        {
+            return IteratorImpl(*array_, static_cast<size_type>(static_cast<difference_type>(index_) + n));
+        }
+        IteratorImpl operator-(const difference_type n) const
+        {
+            return IteratorImpl(*array_, static_cast<size_type>(static_cast<difference_type>(index_) - n));
+        }
+        difference_type operator-(const IteratorImpl& other) const
+        {
+            return static_cast<difference_type>(index_) - static_cast<difference_type>(other.index_);
+        }
+
+        reference operator*()
+        {
+            return this->operator[](0);
+        }
+        const_reference operator*() const
+        {
+            return this->operator[](0);
+        }
+        reference operator[](const difference_type n)
+        {
+            return array_->operator[](static_cast<size_type>(static_cast<difference_type>(index_) + n));
+        }
+        const_reference operator[](const difference_type n) const
+        {
+            return array_->operator[](static_cast<size_type>(static_cast<difference_type>(index_) + n));
+        }
+
+        bool operator==(const IteratorImpl& other) const
+        {
+            return array_ == other.array_ && index_ == other.index_;
+        }
+        bool operator!=(const IteratorImpl& other) const
+        {
+            return !((*this) == other);
+        }
+        bool operator<(const IteratorImpl& other) const
+        {
+            return index_ < other.index_;
+        }
+        bool operator>(const IteratorImpl& other) const
+        {
+            return index_ > other.index_;
+        }
+        bool operator<=(const IteratorImpl& other) const
+        {
+            return index_ <= other.index_;
+        }
+        bool operator>=(const IteratorImpl& other) const
+        {
+            return index_ >= other.index_;
+        }
+
+    private:
+        friend class VariableLengthArray;
+
+        IteratorImpl(A& array, const size_type index) noexcept
+            : array_(&array)
+            , index_(index)
+        {
+        }
+
+        A*        array_ = nullptr;
+        size_type index_ = 0;
+    };
+
+public:
+    using reference       = ReferenceImpl<VariableLengthArray>;
+    using const_reference = ReferenceImpl<const VariableLengthArray>;
+
+    using iterator       = IteratorImpl<VariableLengthArray>;
+    using const_iterator = IteratorImpl<const VariableLengthArray>;
+
+    VariableLengthArray() noexcept
```

Comment:
Should we make the default constructors for VariableLengthArray and this specialization, constexpr?